{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57dba285-c0ef-4a81-9970-6fc341853c07",
   "metadata": {},
   "source": [
    "# Code for figure 2 subpanel C\n",
    "Implementing code here since trying to ensure that I use the same processed data as figure 2 subpanel C."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "032a8e05-040a-42ca-982b-962f504404c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0131060f-4beb-4a46-a714-f838769f98a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_df = pd.read_pickle('./mut_df.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c99c98a6-8916-4274-a1a2-024091c78a2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mut_id</th>\n",
       "      <th>HMDA1-10</th>\n",
       "      <th>HMDA2-1</th>\n",
       "      <th>HMDA2-8</th>\n",
       "      <th>HMDA3-4</th>\n",
       "      <th>HMDA3-5</th>\n",
       "      <th>HMDA3-6</th>\n",
       "      <th>HMDA5-4</th>\n",
       "      <th>HMDA5-5</th>\n",
       "      <th>HMDA5-10</th>\n",
       "      <th>...</th>\n",
       "      <th>GLU3-1</th>\n",
       "      <th>GLU4-1</th>\n",
       "      <th>GLU6-1</th>\n",
       "      <th>GLU7-1</th>\n",
       "      <th>GLU8-1</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>start</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Details</th>\n",
       "      <th>Sequence Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNP-2374-T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2374</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A680S (GCG→TCG)</td>\n",
       "      <td>G→T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNP-20771-A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>20771</td>\n",
       "      <td>insA</td>\n",
       "      <td>intergenic (‑263/+44)</td>\n",
       "      <td>C→A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNP-64352-C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>64352</td>\n",
       "      <td>polB</td>\n",
       "      <td>R477G (CGC→GGC)</td>\n",
       "      <td>G→C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNP-82594-G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>82594</td>\n",
       "      <td>leuA</td>\n",
       "      <td>E312D (GAA→GAC)</td>\n",
       "      <td>T→G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEL-83679-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DEL</td>\n",
       "      <td>83679</td>\n",
       "      <td>leuL</td>\n",
       "      <td>coding (28‑30/87 nt)</td>\n",
       "      <td>(TAG)4→3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 212 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mut_id  HMDA1-10  HMDA2-1  HMDA2-8  HMDA3-4  HMDA3-5  HMDA3-6  \\\n",
       "0   SNP-2374-T         0        0        0        0        0        0   \n",
       "1  SNP-20771-A         0        0        0        0        0        0   \n",
       "2  SNP-64352-C         0        0        0        0        0        0   \n",
       "3  SNP-82594-G         0        0        0        0        0        0   \n",
       "4  DEL-83679-3         0        0        0        0        0        0   \n",
       "\n",
       "   HMDA5-4  HMDA5-5  HMDA5-10  ...  GLU3-1  GLU4-1  GLU6-1  GLU7-1  GLU8-1  \\\n",
       "0        0        0         0  ...       0       0       0       0       0   \n",
       "1        0        0         0  ...       0       0       0       0       0   \n",
       "2        0        0         0  ...       0       0       0       0       0   \n",
       "3        0        0         0  ...       0       0       0       0       0   \n",
       "4        0        0         0  ...       0       0       0       0       0   \n",
       "\n",
       "   mutation_type  start  Gene                Details  Sequence Change  \n",
       "0            SNP   2374  thrA        A680S (GCG→TCG)              G→T  \n",
       "1            SNP  20771  insA  intergenic (‑263/+44)              C→A  \n",
       "2            SNP  64352  polB        R477G (CGC→GGC)              G→C  \n",
       "3            SNP  82594  leuA        E312D (GAA→GAC)              T→G  \n",
       "4            DEL  83679  leuL   coding (28‑30/87 nt)         (TAG)4→3  \n",
       "\n",
       "[5 rows x 212 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "\n",
    "mut_trunc_df = mut_df.copy()\n",
    "\n",
    "with open(\"../Data/Mutation_data/Mutation_protein_changes.json\") as f:\n",
    "    prot_chng_d = json.load(f)\n",
    "mut_trunc_df['Details']=mut_trunc_df['mut_id'].apply(lambda mut_id: prot_chng_d[mut_id])\n",
    "\n",
    "with open(\"../Data/Mutation_data/Mutation_sequence_changes.json\") as f:\n",
    "    mut_seq_chng_d = json.load(f)\n",
    "mut_trunc_df['Sequence Change']=mut_trunc_df['mut_id'].apply(lambda mut_id: mut_seq_chng_d[mut_id])\n",
    "mut_trunc_df.fillna('', inplace=True)\n",
    "\n",
    "\n",
    "mut_trunc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0c4885e3-39d6-4ec1-962d-b11327a18cd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mut_id</th>\n",
       "      <th>HMDA1-10</th>\n",
       "      <th>HMDA2-1</th>\n",
       "      <th>HMDA2-8</th>\n",
       "      <th>HMDA3-4</th>\n",
       "      <th>HMDA3-5</th>\n",
       "      <th>HMDA3-6</th>\n",
       "      <th>HMDA5-4</th>\n",
       "      <th>HMDA5-5</th>\n",
       "      <th>HMDA5-10</th>\n",
       "      <th>...</th>\n",
       "      <th>GLU6-1</th>\n",
       "      <th>GLU7-1</th>\n",
       "      <th>GLU8-1</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>start</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Details</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNP-2374-T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2374</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A680S (GCG→TCG)</td>\n",
       "      <td>G→T</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNP-20771-A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>20771</td>\n",
       "      <td>insA</td>\n",
       "      <td>intergenic (‑263/+44)</td>\n",
       "      <td>C→A</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNP-64352-C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>64352</td>\n",
       "      <td>polB</td>\n",
       "      <td>R477G (CGC→GGC)</td>\n",
       "      <td>G→C</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNP-82594-G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>82594</td>\n",
       "      <td>leuA</td>\n",
       "      <td>E312D (GAA→GAC)</td>\n",
       "      <td>T→G</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEL-83679-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DEL</td>\n",
       "      <td>83679</td>\n",
       "      <td>leuL</td>\n",
       "      <td>coding (28‑30/87 nt)</td>\n",
       "      <td>(TAG)4→3</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 214 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mut_id  HMDA1-10  HMDA2-1  HMDA2-8  HMDA3-4  HMDA3-5  HMDA3-6  \\\n",
       "0   SNP-2374-T         0        0        0        0        0        0   \n",
       "1  SNP-20771-A         0        0        0        0        0        0   \n",
       "2  SNP-64352-C         0        0        0        0        0        0   \n",
       "3  SNP-82594-G         0        0        0        0        0        0   \n",
       "4  DEL-83679-3         0        0        0        0        0        0   \n",
       "\n",
       "   HMDA5-4  HMDA5-5  HMDA5-10  ...  GLU6-1  GLU7-1  GLU8-1  Mutation Type  \\\n",
       "0        0        0         0  ...       0       0       0            SNP   \n",
       "1        0        0         0  ...       0       0       0            SNP   \n",
       "2        0        0         0  ...       0       0       0            SNP   \n",
       "3        0        0         0  ...       0       0       0            SNP   \n",
       "4        0        0         0  ...       0       0       0            DEL   \n",
       "\n",
       "   start  Gene                Details  Sequence Change  mutation size  coding  \n",
       "0   2374  thrA        A680S (GCG→TCG)              G→T              1    True  \n",
       "1  20771  insA  intergenic (‑263/+44)              C→A              1   False  \n",
       "2  64352  polB        R477G (CGC→GGC)              G→C              1    True  \n",
       "3  82594  leuA        E312D (GAA→GAC)              T→G              1    True  \n",
       "4  83679  leuL   coding (28‑30/87 nt)         (TAG)4→3              3    True  \n",
       "\n",
       "[5 rows x 214 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from aledbmutil.mut import get_mut_size, is_coding_mut\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "# refactoring DF to include expected columns\n",
    "mut_trunc_df.rename(columns={\n",
    "    'mutation_type': 'Mutation Type',\n",
    "}, inplace=True)\n",
    "mut_trunc_df.head()\n",
    "\n",
    "mut_trunc_df['mutation size'] = mut_trunc_df.apply(lambda m: get_mut_size(m), axis=1)\n",
    "mut_trunc_df['coding'] = mut_trunc_df.apply(lambda m: is_coding_mut(m.Details) if m.Details != '' else False, axis=1)\n",
    "mut_trunc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c53464aa-e62d-4d21-a268-ebfc1559f8dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mut_id</th>\n",
       "      <th>HMDA1-10</th>\n",
       "      <th>HMDA2-1</th>\n",
       "      <th>HMDA2-8</th>\n",
       "      <th>HMDA3-4</th>\n",
       "      <th>HMDA3-5</th>\n",
       "      <th>HMDA3-6</th>\n",
       "      <th>HMDA5-4</th>\n",
       "      <th>HMDA5-5</th>\n",
       "      <th>HMDA5-10</th>\n",
       "      <th>...</th>\n",
       "      <th>GLU7-1</th>\n",
       "      <th>GLU8-1</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>start</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Details</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>coding</th>\n",
       "      <th>truncation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNP-2374-T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2374</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A680S (GCG→TCG)</td>\n",
       "      <td>G→T</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>non-truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNP-20771-A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>20771</td>\n",
       "      <td>insA</td>\n",
       "      <td>intergenic (‑263/+44)</td>\n",
       "      <td>C→A</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>non-truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNP-64352-C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>64352</td>\n",
       "      <td>polB</td>\n",
       "      <td>R477G (CGC→GGC)</td>\n",
       "      <td>G→C</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>non-truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNP-82594-G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>82594</td>\n",
       "      <td>leuA</td>\n",
       "      <td>E312D (GAA→GAC)</td>\n",
       "      <td>T→G</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>non-truncation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEL-83679-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>DEL</td>\n",
       "      <td>83679</td>\n",
       "      <td>leuL</td>\n",
       "      <td>coding (28‑30/87 nt)</td>\n",
       "      <td>(TAG)4→3</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>non-truncation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 215 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mut_id  HMDA1-10  HMDA2-1  HMDA2-8  HMDA3-4  HMDA3-5  HMDA3-6  \\\n",
       "0   SNP-2374-T         0        0        0        0        0        0   \n",
       "1  SNP-20771-A         0        0        0        0        0        0   \n",
       "2  SNP-64352-C         0        0        0        0        0        0   \n",
       "3  SNP-82594-G         0        0        0        0        0        0   \n",
       "4  DEL-83679-3         0        0        0        0        0        0   \n",
       "\n",
       "   HMDA5-4  HMDA5-5  HMDA5-10  ...  GLU7-1  GLU8-1  Mutation Type  start  \\\n",
       "0        0        0         0  ...       0       0            SNP   2374   \n",
       "1        0        0         0  ...       0       0            SNP  20771   \n",
       "2        0        0         0  ...       0       0            SNP  64352   \n",
       "3        0        0         0  ...       0       0            SNP  82594   \n",
       "4        0        0         0  ...       0       0            DEL  83679   \n",
       "\n",
       "   Gene                Details  Sequence Change  mutation size  coding  \\\n",
       "0  thrA        A680S (GCG→TCG)              G→T              1    True   \n",
       "1  insA  intergenic (‑263/+44)              C→A              1   False   \n",
       "2  polB        R477G (CGC→GGC)              G→C              1    True   \n",
       "3  leuA        E312D (GAA→GAC)              T→G              1    True   \n",
       "4  leuL   coding (28‑30/87 nt)         (TAG)4→3              3    True   \n",
       "\n",
       "       truncation  \n",
       "0  non-truncation  \n",
       "1  non-truncation  \n",
       "2  non-truncation  \n",
       "3  non-truncation  \n",
       "4  non-truncation  \n",
       "\n",
       "[5 rows x 215 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from aledbmutil.mut import is_frameshift, is_disruptive_SNP\n",
    "\n",
    "\n",
    "def _is_gene_truncation(mutation):\n",
    "    # SUB assumed always non-truncation, though could insert a stop codon. Currently don't have logic to find this corner-case.\n",
    "    pred_eff = \"non-truncation\"\n",
    "    \n",
    "    if mutation.coding:\n",
    "        if mutation[\"Mutation Type\"].lower() == \"mob\":\n",
    "            pred_eff = \"truncation\"\n",
    "        if mutation[\"Mutation Type\"].lower() in [\"ins\", \"del\"]:\n",
    "            if is_frameshift(mutation[\"mutation size\"]):\n",
    "                    pred_eff = \"truncation\"\n",
    "        if mutation[\"Mutation Type\"].lower() == \"snp\":\n",
    "            if is_disruptive_SNP(mutation):\n",
    "                pred_eff = \"truncation\"\n",
    "\n",
    "    return pred_eff\n",
    "\n",
    "\n",
    "mut_trunc_df['truncation'] = mut_trunc_df.apply(lambda m: _is_gene_truncation(m), axis=1)\n",
    "mut_trunc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c92fd739-5d4a-42f9-ab81-13e0e19058f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"/bin/bash: -c: line 0: unexpected EOF while looking for matching `''\",\n",
       " '/bin/bash: -c: line 1: syntax error: unexpected end of file']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!!! Don't forget to count the frequency of the presence of a specific mutation across samples so that can accurately capture re-occurring mutations."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
